File: apps/api/plane/app/serializers/scim.py
# Copyright (c) 2023-present Plane Software, Inc. and contributors
# SPDX-License-Identifier: AGPL-3.0-only
# See the LICENSE file for details.

"""Category 11 (docs/feature-specs/11-admin-security-sso.md in
plane-selfhost), feature 2 "SCIM 2.0 natif" - `SCIMToken` serializers for
the ADMIN-facing (workspace-settings) surface, `plane.app.views.workspace.
scim_admin`. This is the "manage my SCIM setup" surface, session-
authenticated, this fork's normal conventions - NOT the `/api/scim/v2/*`
protocol surface itself (`plane.scim`), which never uses DRF serializers
at all (it builds/parses RFC 7644 JSON shapes directly, see
`plane.scim.resources`).

Mirrors `APITokenSerializer`/`APITokenReadSerializer`'s own write/read
split (`plane.app.serializers.api`) - but note `SCIMToken` has no `token`
FIELD to serialize at all (decision #2 - only `token_hash`, HMAC-SHA256,
is ever persisted). `SCIMTokenWriteSerializer.token` is therefore a
`SerializerMethodField` reading a transient `_raw_token` attribute the
CREATE view attaches to the just-created instance in memory (never
persisted) - the one and only place the raw value exists outside the
IdP's own configuration.
"""

from rest_framework import serializers

from plane.db.models import SCIMToken

from .base import BaseSerializer


class SCIMTokenWriteSerializer(BaseSerializer):
    """Creation-response shape only - the raw token is shown exactly
    once. Never used to render a list/detail GET."""

    token = serializers.SerializerMethodField()

    class Meta:
        model = SCIMToken
        fields = ["id", "workspace", "label", "is_active", "last_used_at", "created_at", "created_by", "token"]
        # "token" is a SerializerMethodField (always implicitly read-only,
        # see get_token below) - only the genuine model fields need to be
        # listed here; DRF errors if an explicitly-declared field is also
        # named in read_only_fields.
        read_only_fields = ["id", "workspace", "label", "is_active", "last_used_at", "created_at", "created_by"]

    def get_token(self, obj) -> str:
        return getattr(obj, "_raw_token", "")


class SCIMTokenReadSerializer(BaseSerializer):
    """List/detail shape - `token_hash` is never included, matching
    `APITokenReadSerializer`'s own `exclude = ("token",)` precedent."""

    class Meta:
        model = SCIMToken
        exclude = ("token_hash",)

File: packages/types/src/scim.ts
/**
 * Copyright (c) 2023-present Plane Software, Inc. and contributors
 * SPDX-License-Identifier: AGPL-3.0-only
 * See the LICENSE file for details.
 */

/**
 * Category 11 (docs/feature-specs/11-admin-security-sso.md in
 * plane-selfhost), feature 2 ("SCIM 2.0 natif") - frontend types for the
 * ADMIN-facing "manage my SCIM setup" surface
 * (`plane.app.views.workspace.scim_admin`, apps/api), matching
 * `SCIMTokenReadSerializer`/`SCIMTokenWriteSerializer`
 * (apps/api/plane/app/serializers/scim.py) exactly. Deliberately NOT
 * types for the `/api/scim/v2/*` protocol surface itself (`plane.scim`) -
 * nothing in this frontend checkpoint calls that surface directly, IdPs
 * do.
 *
 * The SCIM provisioning log itself has no dedicated types here - it is a
 * filtered slice of the already-typed `WorkspaceAuditLog`
 * (`TWorkspaceAuditLogListItem`/`TWorkspaceAuditLogDetail`, see
 * packages/types/src/audit.ts), scoped to `SCIM_AUDIT_EVENT_TYPES`
 * (@plane/constants) - reused as-is rather than duplicated.
 */

/** `GET /api/workspaces/<slug>/scim/tokens/` list item - `token_hash` is
 * never included (`SCIMTokenReadSerializer.Meta.exclude = ("token_hash",)`),
 * the raw value is never re-shown after creation. */
export type TSCIMToken = {
  id: string;
  workspace: string;
  label: string;
  is_active: boolean;
  last_used_at: string | null;
  created_at: string;
  created_by: string | null;
};

/** `POST /api/workspaces/<slug>/scim/tokens/` response shape only
 * (`SCIMTokenWriteSerializer`) - `token` is the raw Bearer value, present
 * exactly once, on this response alone. Never returned by the list
 * endpoint above. */
export type TSCIMTokenCreateResponse = TSCIMToken & {
  token: string;
};

export type TSCIMTokenCreatePayload = {
  label?: string;
};
